src/lib/support/BitFlags.h

```diff
@@ -0,0 +1,145 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    Copyright (c) 2013-2017 Nest Labs, Inc.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines functions for manipulating Boolean flags in
+ *      a bitfield.
+ *
+ */
+
+#ifndef CHIP_CHIP_SUPPORT_FLAGUTILS_HPP
+#define CHIP_CHIP_SUPPORT_FLAGUTILS_HPP
+
+#include <stdint.h>
+
+#include <utility>
+
+namespace chip {
+
+/**
+ * Stores bit flags in a type safe manner.
+ *
+ * @tparam StorageType is the underlying storage type (like uint16_t, uint32_t etc.)
+ * @tparam FlagsEnum is the typesafe flags setting
+ */
+template <typename StorageType, typename FlagsEnum>
+class BitFlags
+{
+public:
+    static_assert(sizeof(StorageType) >= sizeof(FlagsEnum), "All flags should fit in the storage type");
```

Comment:
I am unsure if bit operations on signed vs unsiged really result in a difference. Would the things we use like |, &, ~ behave differently on signed/unsigned?

Reply:
I tried in godbolt:

```
enum class E1 :  uint32_t {
  A = 1,
  B = 2,
  C = 4,
};

uint32_t test1() {
    chip::BitFlags<uint32_t, E1> f;

    f.Set(E1::A);
    f.Set(E1::B);
    f.Set(E1::C);

    return f.Raw();
}

enum class E2 {
  A = 1,
  B = 2,
  C = 4,
};

uint32_t test2() {
    chip::BitFlags<int, E2> f;

    f.Set(E2::A);
    f.Set(E2::B);
    f.Set(E2::C);

    return f.Raw();
}
```

both result in the same (super optimized!) code:

```
test1():
        mov     r0, #7
        bx      lr
test2():
        mov     r0, #7
        bx      lr
```

Trying to be more complex:


```
enum class E1 :  uint32_t {
  A = 1,
  B = 2,
  C = 4,
};

uint32_t test1(chip::BitFlags<uint32_t, E1> f) {
    f.Set(E1::A);
    f.Clear(E1::B);
    f.Set(E1::C);

    return f.Raw();
}

enum class E2 {
  A = 1,
  B = 2,
  C = 4,
};

int test2(chip::BitFlags<int, E2> f) {

    f.Set(E2::A);
    f.Clear(E2::B);
    f.Set(E2::C);

    return f.Raw();
}
``` 

gets me

```
test1(chip::BitFlags<unsigned long, E1>):
        bic     r0, r0, #2
        orr     r0, r0, #5
        bx      lr
test2(chip::BitFlags<int, E2>):
        bic     r0, r0, #2
        orr     r0, r0, #5
        bx      lr
```


Based on the above, I believe that we can just not validate types sigdness. It is convenient to just type 'enum class Foo' and not worry that it is an int underneath and there is probably no harm in just allowing 'int32_t' flags if it makes sense for any reason.

Reply:
I was worried about undefined behavior when casting negative values to an unsigned type, but it looks like that's actually defined clearly in terms of what it does, so that's not an issue here.  Given that, probably ok to not worry about it.

Reply:
Does the C spec define this behavior? IIRC, it doesn't and there are some ISAs where sign extension causes 1 bit propagation and some ISAs where it doesn't. Again, IIRC there were differences, for example, in MIPS vs PowerPC in this regard.

Reply:
https://en.cppreference.com/w/c/language/conversion under "Integer conversions" says:

* if the target type can represent the value, the value is unchanged
* otherwise, if the target type is unsigned, the value 2^b, where b is the number of bits in the target type, is repeatedly subtracted or added to the source value until the result fits in the target type. In other words, unsigned integers implement modulo arithmetic. 
* otherwise, if the target type is signed, the behavior is implementation-defined (which may include raising a signal) 

so assigning out-of-range unsigned to signed is bad.  Assigning signed to unsigned is not a problem, per spec; it has well-defined behavior.

We probably _should_ add an assert that `StorageType` is unsigned, still...